File: geocoding/src/main/java/se/arbitur/geocoding/Response.java
package se.arbitur.geocoding;


import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;



public final class Response {
	@SerializedName("results") Result[] results;
	@SerializedName("status") String status;
	@SerializedName("error_message") String errorMessage;


	public Result[] getResults() { return results; }
	public String getStatus() { return status; }
	public String getErrorMessage() { return errorMessage; }

	public Result[] getResults(String addressType) {
		List<Result> filter = new ArrayList<>();

		for (Result result : getResults())
			for (int i = 0; i < result.getAddressTypes().length; i++) {
				String type = result.getAddressTypes()[i];
				if (type.equals(addressType)) {
					filter.add(result);
					break;
				}
			}

		return filter.toArray(new Result[filter.size()]);
	}

}

File: geocoding/src/main/java/se/arbitur/geocoding/Response.kt
package se.arbitur.geocoding


import org.json.JSONObject



class Response(json: JSONObject) {

	val results: Array<Result>?
	val errorMessage: String?
	val status: String = json.getString("status")


	init {
		if (json.has("results")) {
			val results = json.getJSONArray("results")
			this.results = (0 until results.length()).map { Result(results.getJSONObject(it)) }.toTypedArray()
		}
		else
			this.results = null

		if (json.has("error_message"))
			errorMessage = json.getString("error_message")
		else
			this.errorMessage = null
	}


	fun filterResults(type: String): Array<Result>? = results?.filter {
			it.addressTypes.contains(type)
		}!!.toTypedArray()
}